events/handler.go

```diff
@@ -233,15 +240,25 @@ func (self *events) WatchEvents(outChannel chan *Event, request *Request) error
 			"for a call to watch, request.StartTime and request.EndTime must be uninitialized")
 	}
 	newWatcher := newWatch(request, outChannel)
+	self.watcherLock.Lock()
+	defer self.watcherLock.Unlock()
 	self.watchers = append(self.watchers, newWatcher)
 	return nil
 }
 
+func (self *events) updateEventList(e *Event) {
+	self.eventManagerLock.Lock()
+	defer self.eventManagerLock.Unlock()
+	self.eventlist = append(self.eventlist, e)
+}
+
 // method of Events object that adds the argument Event object to the
 // eventlist. It also feeds the event to a set of watch channels
 // held by the manager if it satisfies the request keys of the channels
 func (self *events) AddEvent(e *Event) error {
-	self.eventlist = append(self.eventlist, e)
+	self.updateEventList(e)
+	self.watcherLock.RLock()
+	defer self.watcherLock.RUnlock()
 	for _, watcher := range self.watchers {
 		watchRequest := watcher.request
 		inChannel := watcher.channel
```

Comment:
We should probably not send with the lock (it may take a long time depending on the user) and can lead to deadlocks. Lets create a slice of the channels we're gonna send on (with the lock) and then send to those channels (after dropping the lock).